### ISSUE TITLE
python client get OPENSSL_internal:SSLV3_ALERT_HANDSHAKE_FAILURE

### ISSUE THREAD

**Issue by sshoster** [grpc/grpc#22442]
### What version of gRPC and what language are you using?
grpcio==1.27.2
python 
### What operating system (Linux, Windows,...) and version?
x86_64 x86_64 x86_64 GNU/Linux 
and windows 10 pro

### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.6.8
[root@examples]# pip freeze
certifi==2019.11.28
chardet==3.0.4
grpcio==1.27.2
idna==2.8
oauthlib==3.0.2
protobuf==3.11.3
PyYAML==3.12

### What did you do?
with  python client invoke grpc call with secure channel into ngnix controller 
get the SSLV3_ALERT_HANDSHAKE_FAILURE

while using java client or grpcurl command line works!.

### What did you expect to see?
 HANDSHAKE DONE   in  TLS client

### What did you see instead?

```
I0322 16:34:36.917145785   33378 ev_epoll1_linux.cc:116]     grpc epoll fd: 3
D0322 16:34:36.917261519   33378 ev_posix.cc:174]            Using polling engine: epoll1
D0322 16:34:36.917398916   33378 dns_resolver_ares.cc:503]   Using ares dns resolver
E0322 16:34:36.917444698   33378 trace.cc:65]                Unknown trace var: 'transport_security'
I0322 16:34:36.922114275   33378 ssl_transport_security.cc:222]      HANDSHAKE START -       TLS client start_connect  - !!!!!!
I0322 16:34:36.922187698   33378 ssl_transport_security.cc:222]                 LOOP -    TLS client enter_early_data  - !!!!!!
I0322 16:34:36.922200785   33378 ssl_transport_security.cc:222]                 LOOP -   TLS client read_server_hello  - !!!!!!
E0322 16:34:36.924820031   33378 ssl_transport_security.cc:1378] Handshake failed with fatal error SSL_ERROR_SSL: error:10000410:SSL routines:OPENSSL_internal:SSLV3_ALERT_                    HANDSHAKE_FAILURE.
D0322 16:34:36.924870833   33378 security_handshaker.cc:184] Security handshake failed: {"created":"@1584887676.924855903","description":"Handshake failed","file":"src/cor                    e/lib/security/transport/security_handshaker.cc","file_line":307,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
I0322 16:34:36.924943086   33378 subchannel.cc:1003]         Connect failed: {"created":"@1584887676.924855903","description":"Handshake failed","file":"src/core/lib/secur                    ity/transport/security_handshaker.cc","file_line":307,"tsi_code":10,"tsi_error":"TSI_PROTOCOL_FAILURE"}
I0322 16:34:36.925005405   33378 subchannel.cc:942]          Subchannel 0x1de6e90: Retry in 993 milliseconds
<_InactiveRpcError of RPC that terminated with:
        status = StatusCode.UNAVAILABLE
        details = "failed to connect to all addresses"
        debug_error_string = "{"created":"@1584887676.925001765","description":"Failed to pick subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","                    file_line":3941,"referenced_errors":[{"created":"@1584887676.924996555","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_p                    olicy/pick_first/pick_first.cc","file_line":393,"grpc_status":14}]}"
```

### Anything else we should know about your project / environment?
this is the python code that create the secure channel 
```python
def create_secure_channel(hostname, path, port):
    try:
        certs_folder = Path(path)
        with open(certs_folder / "client.key", 'rb') as fh:
            key = fh.read()
        with open(certs_folder / "client.pem", 'rb') as fh:
            cert = fh.read()
        with open(certs_folder / "server.pem", 'rb') as fh:
            root_ca = fh.read()
        credentials = grpc.ssl_channel_credentials(root_certificates=root_ca, private_key=key,
                                                   certificate_chain=cert)
        #TODO Remove this override after cert fix
        my_server_host_override_string = '172.16.0.1'
        channel = grpc.secure_channel(
            target=hostname + ":" + port,
            credentials=credentials,
            options=(('grpc.ssl_target_name_override', my_server_host_override_string,),))
    except Exception as ex:
        print(ex)
        return ex

    return channel
```

server side using nginx-ingress-controller configuration 
```
server {
                ## SDK endpoints
                listen 9099 ssl http2;
                ssl_certificate                         /etc/ingress-controller/ssl/default-fake-certificate.pem;
                ssl_certificate_key                     /etc/ingress-controller/ssl/default-fake-certificate.pem;
                ssl_prefer_server_ciphers on;
                ssl_certificate_by_lua_block {
                        certificate.call()
                }
                ssl_client_certificate                  /etc/ingress-controller/ssl/ca-default-ca-cert.pem;
                ssl_verify_client                       on;
                ssl_verify_depth                        1;
                ssl_protocols TLSv1.2;
                ## system manager GRPC service with client cert authentication
                location ~ ^/(grpc|input). {
                        grpc_pass              grpc://system-manager-service-grpc:9099;
                        proxy_redirect                          off;
                }
                ## version service REST API placeholder till GRPC is ready
                location ~ ^/versions/(generateId|reportBatch)$ {
                        limit_except POST {
                                deny all;
                        }
                        header_filter_by_lua_block {
                                plugins.run()
                        }
                        body_filter_by_lua_block {
                        }
                        log_by_lua_block {
                           balancer.log()
                                monitor.call()
                                plugins.run()
                        }
                        if ($scheme = https) {
                                more_set_headers                        "Strict-Transport-Security: max-age=15724800; includeSubDomains";
                        }
                        port_in_redirect off;
                        set $balancer_ewma_score -1;
                        set $proxy_upstream_name "default-version-service-80";
                        set $proxy_host          $proxy_upstream_name;
                        set $pass_access_scheme  $scheme;
                        set $pass_server_port    $server_port;
                        set $best_http_host      $http_host;
                        set $pass_port           $pass_server_port;
                        set $proxy_alternative_upstream_name "";
                        client_max_body_size                    0;
                        proxy_set_header Host                   $best_http_host;
                        # Pass the extracted client certificate to the backend
                        # Allow websocket connections
                        proxy_set_header                        Upgrade           $http_upgrade;
                        proxy_set_header                        Connection        $connection_upgrade;
                        proxy_set_header X-Request-ID           $req_id;
                        proxy_set_header X-Real-IP              $remote_addr;
                        proxy_set_header X-Forwarded-For        $remote_addr;
                        proxy_set_header X-Forwarded-Host       $best_http_host;
                        proxy_set_header X-Forwarded-Port       $pass_port;
                        proxy_set_header X-Forwarded-Proto      $pass_access_scheme;
                        proxy_set_header X-Scheme               $pass_access_scheme;
                                                # Pass the original X-Forwarded-For
                        proxy_set_header X-Original-Forwarded-For $http_x_forwarded_for;
                        # mitigate HTTPoxy Vulnerability
                        # https://www.nginx.com/blog/mitigating-the-httpoxy-vulnerability-with-nginx/
                        proxy_set_header Proxy                  "";
                        # Custom headers to proxied server
                        proxy_connect_timeout                   5s;
                        proxy_send_timeout                      60s;
                        proxy_read_timeout                      1800s;
                        proxy_buffering                         off;
                        proxy_buffer_size                       4k;
                        proxy_buffers                           4 4k;
                        proxy_max_temp_file_size                1024m;
                        proxy_request_buffering                 on;
                        proxy_http_version                      1.1;
                        proxy_cookie_domain                     off;
                        proxy_cookie_path                       off;
                        # In case of errors try the next upstream server before returning an error
                        proxy_next_upstream                     error timeout;
                        proxy_next_upstream_timeout             0;
                        proxy_next_upstream_tries               3;
                        proxy_pass http://version-service:80;
                        proxy_redirect                          off;
                }
        }
}
```

**Comment by lidizheng** [comment]
@sshoster There might be some issue in the credential setup, can you post your Java client and grpcurl command as well?

@jiangtaoli2016 Do you have more insight about SSLV3_ALERT_HANDSHAKE_FAILURE? I could not find it in our codebase.

**Comment by jiangtaoli2016** [comment]
It looks like the server side does not set correct cipher suites. 

gRPC client side sets the following cipher suites.
"ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384"

Make sure ngnix controller also sets at least one of the supported suites.

**Comment by sshoster** [comment]
Hi @jiangtaoli2016 
we change both client and server side with the following ciphers and all failed.
**1.)** ECDHE-ECDSA-AES256-GCM-SHA384,ECDHE-RSA-AES256-GCM-SHA384

**2.)** ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:EECDH+AESGCM:EDH+AESGCM:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK

**3.)** EECDH+AESGCM:EDH+AESGCM:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK

all the above give the SSLV3_ALERT_HANDSHAKE_FAILURE

the only one that works was this:
**ALL:!ADH:!EXPORT:!SSLv2:RC4+RSA:+HIGH:+MEDIUM:+LOW**  

can you say why was GRPC client cannot work with advanced ciphers?
please note that i update grpcio to the latest grpcio==1.28.1

**Comment by sshoster** [comment]
update: 
When we change ngnix configuration from 
` ssl_ecdh_curve secp384r1;`
to 
` ssl_ecdh_curve auto;`
Client connection works !!

Can you tell if grpcio supports : secp384r1 ?

if no when will be supported ?

**Comment by jiangtaoli2016** [comment]
@sshoster You are right. I just realized that P256 curve was [hardcoded](https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L798) :(

Let me create a gRPC issue to support 384-bit curves.

**Comment by jiangtaoli2016** [comment]
https://github.com/grpc/grpc/issues/22591 created to keep track. We will fix in Q2.

**Comment by sshoster** [comment]
that's great news thank you @jiangtaoli2016 waiting for that fix :-)

**Comment by sshoster** [comment]
Hi @jiangtaoli2016  ,
can you update on this issue 
thanks

**Comment by jiangtaoli2016** [comment]
I don't have time to fix. Not sure which SSL APIs to call to support multiple curves. 
In the mean time, if you have good idea, please feel free to contribute to gRPC, and I will review.

**Comment by rockspore** [comment]
This should be fixed by #34867. We will support prime256v1, secp384r1 and secp521r1 for TLSv1.2
